patches/tModLoader/Terraria/ModLoader/ModNet.cs

```diff
@@ -225,34 +228,34 @@ internal static bool SyncClientMods(BinaryReader reader, out bool needsReload)
 
 			needsReload = true;
 
-			LocalMod[] localVersions = modFiles.Where(m => m.Name == header.name).ToArray();
-			LocalMod matching = Array.Find(localVersions, mod => header.Matches(mod.modFile));
+			LocalMod matching = modFiles.FirstOrDefault(mod => header.Matches(mod.modFile));
 			if (matching != null) {
-				matching.Enabled = true;
+				MakeEnableOrVersionSwitchExplanationEntries(reloadRequiredExplanationEntries, header, clientMod, matching);
+				if (clientMod == null) {
+					toEnable.Add(matching.Name);
+				}
 				continue;
 			}
 
-			// overwrite an existing version of the mod if there is one
-			if (localVersions.Length > 0)
-				header.path = localVersions[0].modFile.path;
-
-			if (downloadModsFromServers && (header.signed || !onlyDownloadSignedMods))
+			if (downloadModsFromServers && (header.signed || !onlyDownloadSignedMods)) {
 				downloadQueue.Enqueue(header);
-			else
+				MakeDownloadModExplanationEntries(modFiles, reloadRequiredExplanationEntries, header, clientMod);
```

Comment:
Should be `reloadRequiredExplanationEntries.Add(MakeDownloadModExplanation(modFiles, header, clientMod))`